platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/managed/ManagedObjectRegistry.java

```diff
@@ -0,0 +1,220 @@
+public class ManagedObjectRegistry implements AnnotatedServiceLifecycleHandler {
+
+    private final ManagedObjectRegistry parent;
+    private final Map<Class<?>, FactoryRegistration> factories = new ConcurrentHashMap<>();
+
+    public ManagedObjectRegistry(@Nullable ManagedObjectRegistry parent) {
+        this.parent = parent;
+    }
+
+    @Override
+    public List<Class<? extends Annotation>> getAnnotations() {
+        return Collections.singletonList(ManagedObjectProvider.class);
+    }
+
+    @Nullable
+    @Override
+    public Class<? extends Annotation> getImplicitAnnotation() {
+        return null;
+    }
+
+    @Override
+    public void whenRegistered(Class<? extends Annotation> annotation, Registration registration) {
+        assert annotation == ManagedObjectProvider.class;
+        Object instance = registration.getInstance();
+
+        MethodHandles.Lookup lookup = MethodHandles.lookup();
+        for (Method method : instance.getClass().getMethods()) {
+            if (method.isAnnotationPresent(ManagedObjectCreator.class) ||
+                doesParentInterfaceMethodHaveAnnotation(instance.getClass(), method, ManagedObjectCreator.class)
+            ) {
+                if (method.getReturnType() == Void.class) {
+                    throw new IllegalArgumentException("Method " + method + " annotated with @ManagedObjectCreator must return a value");
+                }
+
+                MethodHandle handle;
+                try {
+                     handle = lookup.unreflect(method);
+                } catch (IllegalAccessException e) {
+                    throw UncheckedException.throwAsUncheckedException(e);
+                }
+
+                registerFactory(handle, instance);
+            }
+        }
+    }
+
+    private void registerFactory(MethodHandle handle, Object instance) {
+        validateMethod(handle);
+
+        MethodType type = handle.type();
+        Class<?> returnType = type.returnType();
+        FactoryRegistration existing = factories.put(returnType, new FactoryRegistration(instance, handle));
+
+        if (existing != null) {
+            throw new IllegalArgumentException("Method " + handle + " annotated with @ManagedObjectCreator conflicts with existing factory for type " + returnType);
+        }
+    }
+
+    private static void validateMethod(MethodHandle handle) {
+        MethodType type = handle.type();
+
+        // The 0th parameter is the `this` receiver parameter.
+        int actualParameterCount = type.parameterCount() - 1;
+
+        if (actualParameterCount > 2) {
+            // We only support max 2 arg factories.
+            throw new IllegalArgumentException("Method " + handle + " annotated with @ManagedObjectCreator has too many parameters");
+        }
+
+        for (int i = 0; i < actualParameterCount; i++) {
+            Class<?> parameterType = type.parameterType(i + 1);
+            if (parameterType != Class.class) {
```

Comment:
I don't think this does what I intend it to do. 
This should be an `isAssignableFrom`